docs/src/fake_tensor_and_deferred_init.rst

```diff
@@ -0,0 +1,208 @@
+    ...         self.buf1 = torch.ones([3], device="cpu")
+    ...         self.buf2 = torch.zeros_like(self.buf1)
+
+Assuming we construct ``MyModule`` inside the scope of a ``deferred_init()``
+call with the aforementioned naive approach, both ``buf1`` and ``buf2`` will be
+successfully allocated on the meta device as expected. However when we attempt
+to materialize them, we will hit the problem:
+
+::
+
+    >>> materialize_tensor(my_module.buf1)
+    tensor([1., 1., 1.])
+    >>> materialize_tensor(my_module.buf2)
+    tensor(..., device='meta')
+
+``buf1`` will be successfully materialized on CPU, however ``buf2`` will remain
+on the meta device. The problem is that the implementation of
+``torch.zero_like()`` looks effectively like this:
+
+::
+
+    def zeros_like(src: Tensor):
+        return torch.zeros(src.shape, dtype=src.dtype, device=src.device, ...)
+
+This means when we record the operation in our internal graph the ``device``
+argument that we capture for ``buf2`` will be ``Meta``, not ``CPU``.
+
+Another similar problem happens if the module initialization has some
+device-specific logic:
+
+::
+
+    def foo(self, device: Device) -> Tensor:
+        a = torch.ones([3], device=device)
+
+        return a if a.is_cuda else a + 1
+
+With the naive approach the materialized version of ``a`` will always contain
+``[2., 2., 2.]`` even if the specified real ``device`` was ``CUDA``. This is
+because ``a`` will always be allocated on the meta device and ``is_cuda`` will
+never return ``True``.
+
+In summary in order for materialization to work properly we need a more
+sophisticated approach and this is where the ``Fake`` dispatch key and the fake
+tensor (i.e. ``FakeTensorImpl``) come into play.
+
+Solution
+^^^^^^^^
+``FakeTensorImpl`` is a subclass of ``TensorImpl`` and behaves very similar to
+``OpaqueTensorImpl`` meaning, although it is associated with a real device, it
+has no storage allocated to it. However unlike ``OpaqueTensorImpl`` it also
+holds an internal ``TensorImpl`` that is allocated on the meta backend that acts
+as a "shadow" of the actual tensor.
+
+.. image:: _static/img/fake-tensor.png
+   :alt: FakeTensorImpl
+   :scale: 50%
+   :align: center
+
+The ``Fake`` dispatch key sits in-between Autograd and backend keys where its
+fallback (i.e. catch-all) handler replaces any fake tensor that is passed as an
+argument with its shadow meta tensor and forwards the operation to the meta
+backend. Once the meta backend call returns, it performs the reverse and
+replaces any shadow meta tensor with its fake tensor. Effectively dispatch keys
+above ``Fake`` such as Autograd see fake tensor arguments as regular real
+tensors while dispatch keys below it see them as meta tensors.
+
+.. image:: _static/img/fake-tensor-dispatch.png
+   :alt: Fake Tensor Dispatch
+   :scale: 50%
+   :align: center
+
+Shortcomings
+^^^^^^^^^^^^
+Since internally fake tensors use the meta backend, they have the same
```

Comment:
In the long term, can we just push down these design properties that allow fake tensors to work better with deferred initialization down to meta tensor? And as a result, meta tensor will work just as well for deferred init as fake tensor does today. Or are there significant design blockers?

Reply:
Well, the core team has already ported fake tensor to PyTorch and they prefer it over meta tensor for their future work in TorchDynamo and AOTAutograd. So most likely they will address similar issues in the near future.